core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -106,22 +106,33 @@ class KafkaApis(val requestChannel: RequestChannel,
       case e: FatalExitError => throw e
       case e: Throwable =>
         if (request.requestObj != null) {
-          request.requestObj.handleError(e, requestChannel, request)
-          error("Error when handling request %s".format(request.requestObj), e)
+          def sendResponseCallback(throttleTimeMs: Int) {
+            request.requestObj.handleError(e, requestChannel, request)
+            error("Error when handling request %s".format(request.requestObj), e)
+          }
+          val clientId : String =
+            if (request.requestObj.isInstanceOf[ControlledShutdownRequest])
+              request.requestObj.asInstanceOf[ControlledShutdownRequest].clientId.getOrElse("")
+            else
+              throw new IllegalStateException("Old style requests should only be used for ControlledShutdownRequest")
+          sendResponseMaybeThrottle(request, clientId, sendResponseCallback)
         } else {
-          val response = request.body[AbstractRequest].getErrorResponse(e)
-
-          /* If request doesn't have a default error response, we just close the connection.
-             For example, when produce request has acks set to 0 */
-          if (response == null)
-            requestChannel.closeConnection(request.processor, request)
-          else
-            requestChannel.sendResponse(new Response(request, response))
+          def createResponse(throttleTimeMs: Int): AbstractResponse = {
+            val response = request.body[AbstractRequest].getErrorResponse(throttleTimeMs, e)
+
+            /* If request doesn't have a default error response, we just close the connection.
+               For example, when produce request has acks set to 0 */
+            if (response == null)
+              requestChannel.closeConnection(request.processor, request)
+            response
+          }
+          sendResponseMaybeThrottle(request, createResponse)
```

Comment:
Have moved to a method. I think errors are propagated and logged. We won't close acks=0 connections if `getErrorResponse` threw an exception, but perhaps that is ok since that would still not be a leaked connection?